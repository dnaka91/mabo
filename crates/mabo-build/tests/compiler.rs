use std::{
    fs,
    path::{Path, PathBuf},
};

use insta::{assert_snapshot, glob, with_settings};
use mabo_build::Opts;
use mabo_parser::Schema;

fn strip_path(path: &Path) -> PathBuf {
    path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs"))
        .or_else(|_| path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs_extra")))
        .unwrap()
        .to_owned()
}

#[test]
fn compile_schema() {
    glob!("inputs/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();
        let value = mabo_compiler::simplify_schema(&value);
        let value = mabo_build::compile_schema(&Opts::default(), &value);
        let value = prettyplease::unparse(&syn::parse2(value.clone()).unwrap());

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("compile", value);
        });
    });
}

#[test]
fn compile_schema_extra() {
    glob!("inputs_extra/*.mabo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();
        let value = mabo_compiler::simplify_schema(&value);
        let value = mabo_build::compile_schema(&Opts::default(), &value);
        let value = prettyplease::unparse(&syn::parse2(value.clone()).unwrap());

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("compile_extra", value);
        });
    });
}
