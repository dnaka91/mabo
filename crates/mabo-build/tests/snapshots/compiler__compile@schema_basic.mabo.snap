---
source: crates/mabo-build/tests/compiler.rs
description: "/// Basic struct.\nstruct SampleStruct {\n    a: u32 @1,\n    b: bool @2,\n}\n\n/// Sample enum.\nenum SampleEnum {\n    One @1,\n    Two(u32 @1, u64 @2) @2,\n    Three {\n        field1: u32 @1,\n        field2: bool @2,\n    } @3,\n}"
input_file: crates/mabo-parser/tests/inputs/schema_basic.mabo
---
#[allow(unused_imports)]
use ::mabo::buf::{Decode, Encode, Size};
/// Basic struct.
#[derive(Clone, Debug, PartialEq)]
#[allow(clippy::module_name_repetitions, clippy::option_option)]
pub struct SampleStruct {
    pub a: u32,
    pub b: bool,
}
#[automatically_derived]
impl ::mabo::Encode for SampleStruct {
    #[allow(
        clippy::borrow_deref_ref,
        clippy::deref_addrof,
        clippy::explicit_auto_deref,
        clippy::needless_borrow,
        clippy::too_many_lines,
    )]
    fn encode(&self, w: &mut impl ::mabo::BufMut) {
        let Self { a, b } = self;
        ::mabo::buf::encode_field(
            w,
            ::mabo::FieldId::new(1, ::mabo::FieldEncoding::Varint),
            |w| {
                ::mabo::buf::encode_u32(w, *a);
            },
        );
        ::mabo::buf::encode_field(
            w,
            ::mabo::FieldId::new(2, ::mabo::FieldEncoding::Fixed1),
            |w| {
                ::mabo::buf::encode_bool(w, *b);
            },
        );
        ::mabo::buf::encode_u32(w, ::mabo::buf::END_MARKER);
    }
}
#[automatically_derived]
impl ::mabo::Decode for SampleStruct {
    #[allow(clippy::type_complexity, clippy::too_many_lines)]
    fn decode(r: &mut impl ::mabo::Buf) -> ::mabo::buf::Result<Self> {
        let mut a: Option<u32> = None;
        let mut b: Option<bool> = None;
        loop {
            let id = ::mabo::buf::decode_id(r)?;
            match id.value {
                ::mabo::buf::END_MARKER => break,
                1 => a = Some(::mabo::buf::decode_u32(r)?),
                2 => b = Some(::mabo::buf::decode_bool(r)?),
                _ => ::mabo::buf::decode_skip(r, id.encoding)?,
            }
        }
        Ok(Self {
            a: a
                .ok_or(::mabo::buf::Error::MissingField {
                    id: 1,
                    name: Some("a"),
                })?,
            b: b
                .ok_or(::mabo::buf::Error::MissingField {
                    id: 2,
                    name: Some("b"),
                })?,
        })
    }
}
#[automatically_derived]
impl ::mabo::buf::Size for SampleStruct {
    #[allow(
        clippy::borrow_deref_ref,
        clippy::deref_addrof,
        clippy::explicit_auto_deref,
        clippy::needless_borrow,
        clippy::too_many_lines,
    )]
    fn size(&self) -> usize {
        let Self { a, b } = self;
        ::mabo::buf::size_field(1, || { ::mabo::buf::size_u32(*a) })
            + ::mabo::buf::size_field(2, || { ::mabo::buf::size_bool(*b) })
            + ::mabo::buf::size_u32(::mabo::buf::END_MARKER)
    }
}
/// Sample enum.
#[derive(Clone, Debug, PartialEq)]
#[allow(clippy::module_name_repetitions, clippy::option_option)]
pub enum SampleEnum {
    One,
    Two(u32, u64),
    Three { field1: u32, field2: bool },
}
#[automatically_derived]
impl ::mabo::Encode for SampleEnum {
    #[allow(
        clippy::borrow_deref_ref,
        clippy::deref_addrof,
        clippy::semicolon_if_nothing_returned,
        clippy::too_many_lines,
    )]
    fn encode(&self, w: &mut impl ::mabo::BufMut) {
        match self {
            Self::One => {
                ::mabo::buf::encode_variant_id(w, ::mabo::VariantId::new(1));
            }
            Self::Two(n0, n1) => {
                ::mabo::buf::encode_variant_id(w, ::mabo::VariantId::new(2));
                ::mabo::buf::encode_field(
                    w,
                    ::mabo::FieldId::new(1, ::mabo::FieldEncoding::Varint),
                    |w| {
                        ::mabo::buf::encode_u32(w, *n0);
                    },
                );
                ::mabo::buf::encode_field(
                    w,
                    ::mabo::FieldId::new(2, ::mabo::FieldEncoding::Varint),
                    |w| {
                        ::mabo::buf::encode_u64(w, *n1);
                    },
                );
                ::mabo::buf::encode_u32(w, ::mabo::buf::END_MARKER);
            }
            Self::Three { field1, field2 } => {
                ::mabo::buf::encode_variant_id(w, ::mabo::VariantId::new(3));
                ::mabo::buf::encode_field(
                    w,
                    ::mabo::FieldId::new(1, ::mabo::FieldEncoding::Varint),
                    |w| {
                        ::mabo::buf::encode_u32(w, *field1);
                    },
                );
                ::mabo::buf::encode_field(
                    w,
                    ::mabo::FieldId::new(2, ::mabo::FieldEncoding::Fixed1),
                    |w| {
                        ::mabo::buf::encode_bool(w, *field2);
                    },
                );
                ::mabo::buf::encode_u32(w, ::mabo::buf::END_MARKER);
            }
        }
    }
}
#[automatically_derived]
impl ::mabo::Decode for SampleEnum {
    #[allow(clippy::too_many_lines)]
    fn decode(r: &mut impl ::mabo::Buf) -> ::mabo::buf::Result<Self> {
        match ::mabo::buf::decode_variant_id(r)?.value {
            1 => Ok(Self::One),
            2 => {
                let mut n0: Option<u32> = None;
                let mut n1: Option<u64> = None;
                loop {
                    let id = ::mabo::buf::decode_id(r)?;
                    match id.value {
                        ::mabo::buf::END_MARKER => break,
                        1 => n0 = Some(::mabo::buf::decode_u32(r)?),
                        2 => n1 = Some(::mabo::buf::decode_u64(r)?),
                        _ => ::mabo::buf::decode_skip(r, id.encoding)?,
                    }
                }
                Ok(
                    Self::Two(
                        n0
                            .ok_or(::mabo::buf::Error::MissingField {
                                id: 1,
                                name: None,
                            })?,
                        n1
                            .ok_or(::mabo::buf::Error::MissingField {
                                id: 2,
                                name: None,
                            })?,
                    ),
                )
            }
            3 => {
                let mut field1: Option<u32> = None;
                let mut field2: Option<bool> = None;
                loop {
                    let id = ::mabo::buf::decode_id(r)?;
                    match id.value {
                        ::mabo::buf::END_MARKER => break,
                        1 => field1 = Some(::mabo::buf::decode_u32(r)?),
                        2 => field2 = Some(::mabo::buf::decode_bool(r)?),
                        _ => ::mabo::buf::decode_skip(r, id.encoding)?,
                    }
                }
                Ok(Self::Three {
                    field1: field1
                        .ok_or(::mabo::buf::Error::MissingField {
                            id: 1,
                            name: Some("field1"),
                        })?,
                    field2: field2
                        .ok_or(::mabo::buf::Error::MissingField {
                            id: 2,
                            name: Some("field2"),
                        })?,
                })
            }
            id => Err(::mabo::buf::Error::UnknownVariant(id)),
        }
    }
}
#[automatically_derived]
impl ::mabo::buf::Size for SampleEnum {
    #[allow(
        clippy::borrow_deref_ref,
        clippy::deref_addrof,
        clippy::semicolon_if_nothing_returned,
        clippy::too_many_lines,
    )]
    fn size(&self) -> usize {
        match self {
            Self::One => ::mabo::buf::size_variant_id(1),
            Self::Two(n0, n1) => {
                ::mabo::buf::size_variant_id(2)
                    + ::mabo::buf::size_field(1, || { ::mabo::buf::size_u32(*n0) })
                    + ::mabo::buf::size_field(2, || { ::mabo::buf::size_u64(*n1) })
                    + ::mabo::buf::size_u32(::mabo::buf::END_MARKER)
            }
            Self::Three { field1, field2 } => {
                ::mabo::buf::size_variant_id(3)
                    + ::mabo::buf::size_field(1, || { ::mabo::buf::size_u32(*field1) })
                    + ::mabo::buf::size_field(2, || { ::mabo::buf::size_bool(*field2) })
                    + ::mabo::buf::size_u32(::mabo::buf::END_MARKER)
            }
        }
    }
}

